### DISCUSSION TITLE
Best practices updating useForm initial values

### DISCUSSION THREAD

**Discussion by noahsb** [mantinedev/mantine#2709] [Q&A] [answered]
I’m creating a form with `useForm` and initial values that depend on results from a graphql query. 

I found the documentation on `form.setValues` to help with async data coming in to replace initial values. 

I’m curious if it’s still considered best practice to do something like:

```
  const form = useForm({
    initialValues: {
      name: '',
      email: '',
    },
  });
  const { data } = useQuery(query,
    { onCompleted: (data) => form.setValues(data) })
```

or should form values be updated only with `useEffect`? I’d like to avoid using `useEffect ` where possible. 

Thank you!!

**Comment by rtivital** [comment, marked as answer]
`initialValues` cannot be updated, that why they are called **initial** values. Use `form.setValues` to update values after component was mounted or defer rendering of component that uses `useForm` until you get the data.
